src/server/pfs/server/master.go

```diff
@@ -180,27 +180,52 @@ func (d *driver) finishRepoCommits(ctx context.Context, repoKey string) error {
 				compactor := newCompactor(d.storage, d.env.StorageConfig.StorageCompactionMaxFanIn)
 				taskDoer := d.env.TaskService.NewDoer(StorageTaskNamespace, commit.ID, cache)
 				var totalId *fileset.ID
-				start := time.Now()
-				totalId, err := d.compactCommit(ctx, compactor, taskDoer, commit)
-				if err != nil {
-					return err
-				}
-				details := &pfs.CommitInfo_Details{
-					CompactingTime: types.DurationProto(time.Since(start)),
-				}
-				// Validate the commit.
-				start = time.Now()
-				var validationError string
-				if err := log.LogStep(ctx, "validateCommit", func(ctx context.Context) error {
+				return errors.EnsureStack(d.storage.WithRenewer(ctx, defaultTTL, func(ctx context.Context, renewer *fileset.Renewer) error {
+					start := time.Now()
+					// Compacting the diff before getting the total allows us to compose the
+					// total file set so that it includes the compacted diff.
+					if err := log.LogStep(ctx, "compactDiffFileSet", func(ctx context.Context) error {
+						return d.compactDiffFileSet(ctx, compactor, taskDoer, renewer, commit)
+					}); err != nil {
+						return err
+					}
 					var err error
-					validationError, details.SizeBytes, err = compactor.Validate(ctx, taskDoer, *totalId)
-					return err
-				}); err != nil {
-					return err
-				}
-				details.ValidatingTime = types.DurationProto(time.Since(start))
-				// Finish the commit.
-				return d.finalizeCommit(ctx, commit, validationError, details, totalId)
+					// A commit may have a compacted total file set, but an uncompacted diff file set.
+					// If a total exists, this commit has been through total file set compaction before
+					// and the rest of the compaction steps can be skipped.
+					totalId, err = d.commitStore.GetTotalFileSet(ctx, commit)
```

Comment:
I think both of these variables can be declared here, and the check for a non-nil error can probably be removed. We only need to handle the nil case, and proceed otherwise.